lib/trino-parquet/src/main/java/io/trino/parquet/ValidationHash.java

```diff
@@ -0,0 +1,144 @@
+                    ValidationHash.class,
+                    "mapHash",
+                    MethodType.methodType(long.class, Type.class, ValidationHash.class, ValidationHash.class, Block.class, int.class));
+            ARRAY_HASH = lookup().findStatic(
+                    ValidationHash.class,
+                    "arrayHash",
+                    MethodType.methodType(long.class, Type.class, ValidationHash.class, Block.class, int.class));
+            ROW_HASH = lookup().findStatic(
+                    ValidationHash.class,
+                    "rowHash",
+                    MethodType.methodType(long.class, Type.class, ValidationHash[].class, Block.class, int.class));
+        }
+        catch (Exception e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    // This should really come from the environment, but there is not good way to get a value here
+    private static final TypeOperators VALIDATION_TYPE_OPERATORS_CACHE = new TypeOperators();
+
+    public static ValidationHash createValidationHash(Type type)
+    {
+        requireNonNull(type, "type is null");
+        if (type.getTypeSignature().getBase().equals(MAP)) {
+            ValidationHash keyHash = createValidationHash(type.getTypeParameters().get(0));
+            ValidationHash valueHash = createValidationHash(type.getTypeParameters().get(1));
+            return new ValidationHash(MAP_HASH.bindTo(type).bindTo(keyHash).bindTo(valueHash));
+        }
+
+        if (type.getTypeSignature().getBase().equals(ARRAY)) {
+            ValidationHash elementHash = createValidationHash(type.getTypeParameters().get(0));
+            return new ValidationHash(ARRAY_HASH.bindTo(type).bindTo(elementHash));
+        }
+
+        if (type.getTypeSignature().getBase().equals(ROW)) {
+            ValidationHash[] fieldHashes = type.getTypeParameters().stream()
+                    .map(ValidationHash::createValidationHash)
+                    .toArray(ValidationHash[]::new);
+            return new ValidationHash(ROW_HASH.bindTo(type).bindTo(fieldHashes));
+        }
+
+        return new ValidationHash(VALIDATION_TYPE_OPERATORS_CACHE.getHashCodeOperator(type, InvocationConvention.simpleConvention(FAIL_ON_NULL, BLOCK_POSITION)));
+    }
+
+    private final MethodHandle hashCodeOperator;
+
+    private ValidationHash(MethodHandle hashCodeOperator)
+    {
+        this.hashCodeOperator = requireNonNull(hashCodeOperator, "hashCodeOperator is null");
+    }
+
+    public long hash(Block block, int position)
+    {
+        if (block.isNull(position)) {
+            return NULL_HASH_CODE;
+        }
+        try {
+            return (long) hashCodeOperator.invokeExact(block, position);
+        }
+        catch (Throwable throwable) {
+            throwIfUnchecked(throwable);
+            throw new RuntimeException(throwable);
+        }
+    }
+
+    private static long mapHash(Type type, ValidationHash keyHash, ValidationHash valueHash, Block block, int position)
```

Comment:
I'm not sure why there is even a `isNull` check in `mapSkipNullKeysHash` given that the null case is already handled by `hash` method.
Maybe @dain knows the reason behind that ?
The xor was avoided to avoid having a function which gives the same result if the key and value are flipped.